File: packages/sdk-ts/src/resources/passports.ts
import type { HttpClient } from '../http.js';

// ─── Types ───────────────────────────────────────────────────────────────────

export interface IssuePassportParams {
  agentId: string;
  grantId: string;
  allowedMPPCategories: string[];
  maxTransactionAmount: { amount: number; currency: string };
  paymentRails?: string[];
  expiresIn?: string;
  parentPassportId?: string;
}

export interface IssuedPassportResponse {
  passportId: string;
  credential: Record<string, unknown>;
  encodedCredential: string;
  expiresAt: string;
}

export interface GetPassportResponse {
  status: string;
  [key: string]: unknown;
}

export interface RevokePassportResponse {
  revoked: boolean;
  revokedAt: string;
}

export interface ListPassportsParams {
  agentId?: string;
  grantId?: string;
  status?: string;
}

// ─── Client ──────────────────────────────────────────────────────────────────

export class PassportsClient {
  readonly #http: HttpClient;

  constructor(http: HttpClient) {
    this.#http = http;
  }

  issue(params: IssuePassportParams): Promise<IssuedPassportResponse> {
    return this.#http.post<IssuedPassportResponse>('/v1/passport/issue', params);
  }

  get(passportId: string): Promise<GetPassportResponse> {
    return this.#http.get<GetPassportResponse>(
      `/v1/passport/${encodeURIComponent(passportId)}`,
    );
  }

  revoke(passportId: string): Promise<RevokePassportResponse> {
    return this.#http.post<RevokePassportResponse>(
      `/v1/passport/${encodeURIComponent(passportId)}/revoke`,
    );
  }

  list(params?: ListPassportsParams): Promise<IssuedPassportResponse[]> {
    const searchParams = new URLSearchParams();
    if (params) {
      for (const [key, value] of Object.entries(params)) {
        if (value !== undefined) searchParams.set(key, String(value));
      }
    }
    const qs = searchParams.toString();
    return this.#http.get<IssuedPassportResponse[]>(
      `/v1/passports${qs ? `?${qs}` : ''}`,
    );
  }
}

File: packages/sdk-py/src/grantex/resources/_passports.py
from __future__ import annotations

from .._http import HttpClient
from .._types import (
    GetPassportResponse,
    IssuedPassportResponse,
    IssuePassportParams,
    ListPassportsParams,
    ListPassportsResponse,
    RevokePassportResponse,
)


class PassportsClient:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def issue(self, params: IssuePassportParams) -> IssuedPassportResponse:
        data = self._http.post("/v1/passport/issue", params.to_dict())
        return IssuedPassportResponse.from_dict(data)

    def get(self, passport_id: str) -> GetPassportResponse:
        from urllib.parse import quote

        data = self._http.get(f"/v1/passport/{quote(passport_id, safe='')}")
        return GetPassportResponse.from_dict(data)

    def revoke(self, passport_id: str) -> RevokePassportResponse:
        from urllib.parse import quote

        data = self._http.post(
            f"/v1/passport/{quote(passport_id, safe='')}/revoke"
        )
        return RevokePassportResponse.from_dict(data)

    def list(
        self, params: ListPassportsParams | None = None
    ) -> ListPassportsResponse:
        path = "/v1/passports"
        if params is not None:
            query = params.to_query()
            if query:
                from urllib.parse import urlencode

                path = f"{path}?{urlencode(query)}"
        data = self._http.get(path)
        # Server returns a bare JSON array
        return ListPassportsResponse.from_list(data)
